### ISSUE TITLE
[tm_json.h] unused function warning

### ISSUE THREAD

**Issue by fluffrabbit** [to-miz/tm#2]
Great project. There is still a warning if `TM_JSON_IMPLEMENTATION` is defined (both GCC and Clang):

```
g++ -std=c++11 -Wall -Wpedantic -O2 -o tmj_example tmj_example.cpp
In file included from tmj_example.cpp:2:0:
tm_json.h:819:16: warning: ‘tm_bool stringEqualsIgnoreCaseNull(const char*, size_t, const char*)’ defined but not used [-Wunused-function]
 static tm_bool stringEqualsIgnoreCaseNull(const char* a, size_t aSize, const char* b) {
                ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

<s>I have attached the start of an example project. I did get things working before, but since the update my code stopped working, and absent documentation I don't feel like going all Indiana Jones on it in my present state. Many thanks for even small insight into how the library works.</s>

EDIT: It was my code, not yours. The library still works the same way. Still, it would be nice to have some information about (for example) how to access a yet-unknown object as key/value pairs.

**Comment by to-miz** [comment]
Seems that I actually did introduce a new bug with the last version. I will push a new version soon, but this time with more thorough testing. I will also include example code to showcase how to use it.

Note that using `jsonAllocateDocumentEx` with `JSON_READER_ALL` will enable **ALL** extensions, so the parser will accept input, that is no longer strictly speaking json. You can use the ex version for example to parse [json5](https://json5.org/). Most of the time it is enough to use `jsonAllocateDocument` with the flags `JSON_READER_SINGLE_LINE_COMMENTS | JSON_READER_TRAILING_COMMA`. This will allow `// single line comments` and trailing commas as in `[1,2,]`, which are very nice to have when handwriting json.

I think you meant this snippet in your example to iterate over an array of objects/nodes:
`for (auto node : exampleJsonValue.getObjectArray())`
Note that I simply used `auto` instead of `auto&`, this library is designed in a way, where you are encouraged to take things by value. In this case `auto` means `JsonObject` and `getObjectArray()` is just a shorthand of this snippet:
```
for (JsonValue value : exampleJsonValue.getArray()) {
   JsonObject node = value.getObject();
   /* Code using node... */
}
```

`getObjectArray` will treat a given `JsonValue` as if was an array of `JsonObject`s.

As to printing `JsonStringView` using `std::cout`, you can utilize the customization point of ostreams like this:
```
std::ostream& operator<<(std::ostream& o, JsonStringView str) {
	o.write(str.data, str.size);
	return o;
}
```
Then you can just write `std::cout << node.name;` and it will just work.

**Comment by fluffrabbit** [comment]
Happy 2019!

It was an invalid token error I was getting, so I assumed you had added an additional validation check. Since I was loading JSON generated by Tiled, I figured the parser needed to be looser than a barn door.

I have not seen the benefit of using `JsonObject` over `JsonValue`, as they both share the same properties, and adding a property to the end of a variable doesn't bother me. The structs `JsonArray` and `JsonObjectArray` are likewise very similar, and it's easy to get confused about these things. When you throw `JsonNode`s into the mix it scambles it more. This is adult stuff here, not the kind of programming I started with.

"Just" using `ostream`s is a bit above my pay grade, but it's neat to see another way of doing things.

Before the next release, could you try compiling with Clang? I noticed this additional warning:

```
tm_json.h:2919:37: warning: suggest braces around initialization of subobject [-Wmissing-braces]
    JsonValue result = {JVAL_NULL, {TM_NULL, 0}};
                                    ^~~~~~~~~~
                                    {         }
tm_json.h:242:25: note: expanded from macro 'TM_NULL'
        #define TM_NULL nullptr
                        ^
```

**Comment by to-miz** [comment]
Compilation errors should be fixed with 585c58d78ac04791eaa55332ebac4fd0386606f8. I checked with gcc7.3.0, clang7.0.0 and msvc2017.
The bug I introduced with the last push is also fixed (Keyword parsing was broken, something like `"key": null`).

**Comment by fluffrabbit** [comment]
In some places errorType has become error.type, but I'm also seeing errorType. Whatever. It can now read Tiled maps in the strictest mode, and no compiler wawnings in GCC or Clang. Hooray! It feels stable now. No complaints here (so far).
